src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Automation/Peers/DataGridAutomationPeer.cs

```diff
@@ -137,15 +138,45 @@ int IGridProvider.ColumnCount
         {
             get
             {
-                return this.OwningDataGrid.Columns.Count;
+                int visibleColumnCount = 0;
+
+                foreach (var col in OwningDataGrid.Columns)
+                {
+                    if (((Visibility)col.GetValue(DataGridColumn.VisibilityProperty)) == Visibility.Visible)
+                    {
+                        visibleColumnCount++;
+                    }
+                }
+
+                return visibleColumnCount;
             }
         }
 
         int IGridProvider.RowCount
         {
             get
             {
-                return this.OwningDataGrid.Items.Count;
+                int rowCount = this.OwningDataGrid.Items.Count;
+
+                var editableItems = (IEditableCollectionView)this.OwningDataGrid.Items;
+
+                if (editableItems.NewItemPlaceholderPosition != NewItemPlaceholderPosition.None)
+                {
+                    --rowCount;
+                }
+
+                for (int i = 0; i < rowCount; i++)
```

Comment:
Bad assumption here, Need to do different things depending on whether the placeholder is at the beginning or the end.  It's as easy as declaring i and incrementing for the front case.  Just use a switch and go from there.

Reply:
We can't ignore the placeholder, it's navigable.  However, if a dev sets placeholder to none, the datagrid overrides this choice (even if it doesn't show the placeholder) and the rowcount will be +1 over navigable rows.  So when it's there we should count it, when it's not we should decrement.

Reply:
Actually, the placeholder is collapsed when it should be, Generally, we should just be fine not counting it based on visibility.